compiler/rustc_mir_transform/src/value_numbering.rs

```diff
@@ -0,0 +1,907 @@
+                    }
+                }
+                VnOperand::Constant(_) => {}
+            }
+        }
+
+        for vn_place_elem in vn_place.projection.iter_mut() {
+            modified = self.substitute_values_in_vn_place_elem(vn_place_elem) || modified;
+        }
+        modified
+    }
+
+    fn substitute_values_in_vn_place_elem(&self, vn_place_elem: &mut VnPlaceElem<'tcx>) -> bool {
+        match *vn_place_elem {
+            ProjectionElem::Index(ref mut orig_vn_local) => {
+                if let Some(vn_operand) = self.values.get(&VnRvalue::use_of_local(orig_vn_local)) {
+                    if let Some(vn_local) = vn_operand.get_single_local() {
+                        if self.local_may_be_materialized_here(&vn_local) {
+                            *orig_vn_local = vn_local;
+                            return true;
+                        }
+                    }
+                }
+            }
+            ProjectionElem::Deref
+            | ProjectionElem::ConstantIndex { .. }
+            | ProjectionElem::Subslice { .. }
+            | ProjectionElem::Downcast(_, _)
+            | ProjectionElem::Field(_, _) => {}
+        }
+        false
+    }
+
+    fn substitute_values_in_vn_operand(&self, vn_operand: &mut VnOperand<'tcx>) -> bool {
+        let mut modified = false;
+        if let Some(vn_new_operand) = self.values.get(&VnRvalue::Use((*vn_operand).clone())) {
+            if self.operand_may_be_materialized_here(vn_new_operand) {
+                *vn_operand = (*vn_new_operand).clone();
+                modified = true;
+            }
+        }
+        modified
+    }
+
+    // Directly performs copy/constant propagation optimizations in the given MIR Operand.
+    fn substitute_values_in_operand(&self, operand: &mut Operand<'tcx>) -> bool {
+        if let Some(mut vn_operand) = self.vn_operand_for(operand) {
+            if self.substitute_values_in_vn_operand(&mut vn_operand) {
+                *operand = vn_operand.to_operand(self.tcx);
+                return true;
+            }
+        }
+        false
+    }
+
+    // Returns true if the given VnPlace describes an immutable location. Immutable locations are
+    // guaranteed to be immutable for the life of the place.
+    fn vn_place_is_immutable(&self, place: &VnPlace<'tcx>, local_decls: &LocalDecls<'tcx>) -> bool {
+        debug!("vn_place_is_immutable({:?})", place);
+        let local_decl = &local_decls[place.local.local];
+        if let Mutability::Mut = local_decl.mutability {
+            return false;
+        }
+
+        let mut ty_kind = local_decl.ty.kind();
+        for place_elem in &place.projection {
+            match *place_elem {
+                ProjectionElem::Deref => match ty_kind {
+                    ty::Ref(_, inner_ty, Mutability::Not) => {
+                        ty_kind = inner_ty.kind();
+                    }
+                    _ => return false,
+                },
```

Comment:
Place behind a shared reference might mutable?

```rust
fn f(a: *const u32, b: *mut u32) -> u32 {
    let a = unsafe {&*a};
    let c = *a;
    unsafe {*b = 2};
    c
}

fn main() {
    let mut x = 1;
    let p = std::ptr::addr_of_mut!(x);
    assert_eq!(f(p, p), 1);
}
```

```console
$ rustc +stage1 -O a.rs && ./a
$ rustc +stage1 -O -Znumber_values a.rs && ./a
thread 'main' panicked at 'assertion failed: `(left == right)`
  left: `2`,
 right: `1`', a.rs:11:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

Reply:
This is because of NLL I guess. I suppose we should check to see whether the live ranges of the immutable reference overlap the mutation site. If they don't we have to kill the immutable reference.

Reply:
OK, new plan: contents of an immutable reference are considered immutable if and only if the immutable reference either came from a function parameter or has a use reachable from the current basic block. I *think* this is sound. It would probably be possible to intertwine this pass with NLL data structures, but I'm very hesitant to do that.

Reply:
In other words, the reference must not only be immutable, it must be *immutably borrowed across* the entire basic block.

Here's a modified form of the example that fails to borrow check because it fails the "uses reachable from the statement" check: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=2dcd401a1b807854207c732e2153fc3a

And here's an example that shows that the "function parameter" check is sound because the borrow check rejects an attempt to violate it: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=2dcd401a1b807854207c732e2153fc3a

Reply:
Note that the optimized example fails the borrow check: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=f1a5acf2c7389f89150d9e25f6d1002c